Entering login screen might result in a crash

Under some conditions entering login screen might result in an application crash caused by network call made to authenticate in userless mode. Requirement is to be not logged in, have no valid authentication data stored, and then enter login screen. Sadly I didn't manage to reproduce this issue as going through earlier screen automatically prevents this case from being reproducible (JrawReddit is being initialized as singleton earlier in background thread).

I've attached a potential fix but I don't want to commit it unless we manage to reproduce the issue and verify that the fix doesn't cause different hidden problems.

https://github.com/Tunous/Dank/blob/0d4226a9122123187de91610b4e51f5a5fadb43b/app/src/main/java/me/saket/dank/reddit/jraw/JrawReddit.kt#L38

<details>
<summary>Stacktrace</summary>

```
java.lang.RuntimeException: 
  at android.app.ActivityThread.performLaunchActivity (ActivityThread.java:2473)
  at android.app.ActivityThread.handleLaunchActivity (ActivityThread.java:2535)
  at android.app.ActivityThread.access$1100 (ActivityThread.java:154)
  at android.app.ActivityThread$H.handleMessage (ActivityThread.java:1396)
  at android.os.Handler.dispatchMessage (Handler.java:102)
  at android.os.Looper.loop (Looper.java:148)
  at android.app.ActivityThread.main (ActivityThread.java:5582)
  at java.lang.reflect.Method.invoke (Native Method)
  at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run (ZygoteInit.java:726)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:616)

Caused by: android.os.NetworkOnMainThreadException: 
  at android.os.StrictMode$AndroidBlockGuardPolicy.onNetwork (StrictMode.java:1273)
  at java.net.InetAddress.lookupHostByName (InetAddress.java:431)
  at java.net.InetAddress.getAllByNameImpl (InetAddress.java:252)
  at java.net.InetAddress.getAllByName (InetAddress.java:215)
  at okhttp3.Dns$1.lookup (Dns.java:40)
  at okhttp3.internal.connection.RouteSelector.resetNextInetSocketAddress (RouteSelector.java:185)
  at okhttp3.internal.connection.RouteSelector.nextProxy (RouteSelector.java:149)
  at okhttp3.internal.connection.RouteSelector.next (RouteSelector.java:84)
  at okhttp3.internal.connection.StreamAllocation.findConnection (StreamAllocation.java:214)
  at okhttp3.internal.connection.StreamAllocation.findHealthyConnection (StreamAllocation.java:135)
  at okhttp3.internal.connection.StreamAllocation.newStream (StreamAllocation.java:114)
  at okhttp3.internal.connection.ConnectInterceptor.intercept (ConnectInterceptor.java:42)
  at okhttp3.internal.http.RealInterceptorChain.proceed (RealInterceptorChain.java:147)
  at okhttp3.internal.http.RealInterceptorChain.proceed (RealInterceptorChain.java:121)
  at okhttp3.internal.cache.CacheInterceptor.intercept (CacheInterceptor.java:93)
  at okhttp3.internal.http.RealInterceptorChain.proceed (RealInterceptorChain.java:147)
  at okhttp3.internal.http.RealInterceptorChain.proceed (RealInterceptorChain.java:121)
  at okhttp3.internal.http.BridgeInterceptor.intercept (BridgeInterceptor.java:93)
  at okhttp3.internal.http.RealInterceptorChain.proceed (RealInterceptorChain.java:147)
  at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept (RetryAndFollowUpInterceptor.java:126)
  at okhttp3.internal.http.RealInterceptorChain.proceed (RealInterceptorChain.java:147)
  at okhttp3.internal.http.RealInterceptorChain.proceed (RealInterceptorChain.java:121)
  at okhttp3.RealCall.getResponseWithInterceptorChain (RealCall.java:200)
  at okhttp3.RealCall.execute (RealCall.java:77)
  at net.dean.jraw.http.OkHttpNetworkAdapter.execute (OkHttpNetworkAdapter.kt:16)
  at net.dean.jraw.oauth.OAuthHelper.applicationOnlyOAuthData$lib (OAuthHelper.kt:88)
  at net.dean.jraw.oauth.OAuthHelper.automatic (OAuthHelper.kt:30)
  at net.dean.jraw.oauth.AccountHelper.switchToUserless (AccountHelper.kt:74)
  at me.saket.dank.reddit.jraw.JrawReddit.<init> (JrawReddit.kt:38)
  at me.saket.dank.reddit.jraw.JrawReddit_Factory.get (JrawReddit_Factory.java:38)
  at me.saket.dank.reddit.jraw.JrawReddit_Factory.get (JrawReddit_Factory.java:12)
  at me.saket.dank.reddit.RedditModule_RedditFactory.get (RedditModule_RedditFactory.java:22)
  at me.saket.dank.reddit.RedditModule_RedditFactory.get (RedditModule_RedditFactory.java:9)
  at dagger.internal.DoubleCheck.get (DoubleCheck.java:47)
  at me.saket.dank.ui.authentication.LoginActivity.onCreate (LoginActivity.java:104)
  at android.app.Activity.performCreate (Activity.java:6309)
  at android.app.Instrumentation.callActivityOnCreate (Instrumentation.java:1108)
  at android.app.ActivityThread.performLaunchActivity (ActivityThread.java:2426)
```
</details>

<details>
<summary>Potential fix</summary>

```diff
Index: app/src/main/java/me/saket/dank/reddit/jraw/JrawReddit.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- app/src/main/java/me/saket/dank/reddit/jraw/JrawReddit.kt	(revision 219a5b71b62eb8ab92be30aef8368980bb585ae9)
+++ app/src/main/java/me/saket/dank/reddit/jraw/JrawReddit.kt	(date 1568132381364)
@@ -1,5 +1,7 @@
 package me.saket.dank.reddit.jraw
 
+import io.reactivex.Completable
+import io.reactivex.schedulers.Schedulers.io
 import io.reactivex.subjects.BehaviorSubject
 import me.saket.dank.reddit.Reddit
 import me.saket.dank.ui.user.UserSessionRepository
@@ -35,7 +37,10 @@
       }
       else -> {
         Timber.i("Switching to userless")
-        accountHelper.switchToUserless()
+        Completable.fromAction {
+          accountHelper.switchToUserless()
+        }.subscribeOn(io())
+          .subscribe()
       }
     }
 ```
</details>

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,6 +7,9 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 ### Changed
 - Renamed application to "Dawn" ([#74](https://github.com/Tunous/Dawn/pull/74))
 
+### Fixed
+- Occasional crash when entering login screen ([#103](https://github.com/Tunous/Dawn/pull/103))
+
 ## [0.8.0] - 2019-08-22
 ### Added
 - Subscribing to /r/all ([#28](https://github.com/Tunous/Dank/pull/28))

--- a/app/src/main/java/me/saket/dank/reddit/jraw/JrawReddit.kt
+++ b/app/src/main/java/me/saket/dank/reddit/jraw/JrawReddit.kt
@@ -1,5 +1,7 @@
 package me.saket.dank.reddit.jraw
 
+import io.reactivex.Completable
+import io.reactivex.schedulers.Schedulers.io
 import io.reactivex.subjects.BehaviorSubject
 import me.saket.dank.reddit.Reddit
 import me.saket.dank.ui.user.UserSessionRepository
@@ -35,7 +37,10 @@ class JrawReddit @Inject constructor(
       }
       else -> {
         Timber.i("Switching to userless")
-        accountHelper.switchToUserless()
+        Completable.fromAction {
+          accountHelper.switchToUserless()
+        }.subscribeOn(io())
+          .subscribe()
       }
     }
 

